crates/zkvm/zkvm/src/errors.rs

```diff
@@ -1,103 +1,145 @@
 use thiserror::Error;
 
+use crate::{ProofType, ZkVm};
+
+/// A convenient alias for results in the ZkVM.
 pub type ZkVmResult<T> = Result<T, ZkVmError>;
 
-/// Represents different types of errors that can occur in the ZkVM system
 #[derive(Debug, Error)]
 pub enum ZkVmError {
-    /// Error during serialization of input or output
-    #[error("Serialization failed")]
-    SerializationError {
-        /// Specific serialization error type
-        #[source]
-        source: SerializationErrorSource,
-    },
-
-    /// Error during deserialization of input or output
-    #[error("Deserialization failed")]
-    DeserializationError {
-        /// Specific deserialization error type
-        #[source]
-        source: DeserializationErrorSource,
-    },
-    /// Error during proof generation
-    #[error("Proof generation failed")]
+    #[error("Proof generation failed: {0}")]
     ProofGenerationError(String),
 
-    /// Error during proof verification
-    #[error("Proof verification failed")]
+    #[error("Proof verification failed: {0}")]
     ProofVerificationError(String),
 
-    /// Input-related errors
-    #[error("Input validation failed")]
-    InputError(String),
+    #[error("Input validation failed: {0}")]
+    InvalidInput(#[from] ZkVmInputError),
 
-    /// ELF-related errors
-    #[error("ELF validation failed")]
+    #[error("ELF validation failed: {0}")]
     InvalidELF(String),
 
-    /// Verification Key related errors
     #[error("Invalid Verification Key")]
-    InvalidVerificationKey,
+    InvalidVerificationKey(#[from] ZkVmVerificationKeyError),
 
-    /// Generic error for other cases
-    #[error("An unexpected error occurred")]
+    #[error("Invalid proof receipt")]
+    InvalidProofReceipt(#[from] ZkVmProofError),
+
+    #[error("Output extraction failed")]
+    OutputExtractionError {
+        #[source]
+        source: DataFormatError,
+    },
+
+    #[error("An unexpected error occurred: {0}")]
     Other(String),
 }
 
-/// Enum to statically handle different serialization error sources
 #[derive(Debug, Error)]
-pub enum SerializationErrorSource {
-    /// Bincode serialization error
-    #[error("Bincode serialization error")]
-    Bincode(#[from] bincode::Error),
+pub enum DataFormatError {
+    #[error("error using bincode: {source}")]
+    Bincode {
+        #[source]
+        source: bincode::Error,
+    },
 
-    /// Borsh serialization error
-    #[error("Borsh serialization error")]
-    Borsh(#[from] borsh::io::Error),
+    #[error("error using borsh: {source}")]
+    Borsh {
+        #[source]
+        source: borsh::io::Error,
+    },
 
-    /// Serde serialization error
-    #[error("Serde serialization error: {0}")]
+    #[error("error using serde: {0}")]
```

Comment:
Can take out the `error using ` prefixes on all these error strings, it's implied.